### PR TITLE
daemon: switch to 'ensure' workflow for AppArmor profiles

### PR DESCRIPTION
**Pull request by cyphar**
In certain cases (unattended upgrades), system services can disable
loaded AppArmor profiles. However, since /etc being read-only is a
supported setup we cannot just write a copy of the profile to
/etc/apparmor.d.

Instead, dynamically load the docker-default AppArmor profile if a
container is started with that profile set. This code will short-cut if
the profile is already loaded.

cute bunneh

Fixes: #29097
Fixes: 2f7596a ("apparmor: do not save profile to /etc/apparmor.d")
Signed-off-by: Aleksa Sarai <asarai@suse.de>

### CODE REVIEW

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Comment by cyphar** [comment]
Note that my test case for this was:


```
% docker run --rm -it alpine echo "hello"
hello
% apparmor_parser -R /etc/apparmor.d/docker
Warning failed to create cache: (null)
% docker run --rm -it alpine echo "hello"
hello
```

Which I think is sufficient to check that the reloading works as expected.

File: daemon/oci_linux.go [hunk 1]
```diff
@@ -733,11 +733,19 @@ func (daemon *Daemon) createSpec(c *container.Container) (*specs.Spec, error) {
...
+		var appArmorProfile string
+		if c.AppArmorProfile != "" {
 			appArmorProfile = c.AppArmorProfile
 		} else if c.HostConfig.Privileged {
 			appArmorProfile = "unconfined"
+		} else {
+			// Unattended upgrades and other fun services can unload AppArmor
+			// profiles inadvertently. So in order to make sure that we keep
+			// the profile enabled, we will dynamically reload it if necessary.
+			if err := ensureDefaultAppArmorProfile(); err != nil {
```
Comment (tonistiigi) [inline daemon/oci_linux.go]: Do this later with ` if appArmorProfile == defaultApparmorProfile` condition to do it also loaded if `c.AppArmorProfile` equals `docker-default`.

Comment (cyphar) [inline daemon/oci_linux.go]: Fixed, thanks.

**Comment by tonistiigi** [comment]
LGTM

cc @justincormack

**Review by justincormack — APPROVED** [review]

**Comment by justincormack** [comment]
I guess there is a very short race still, but that seems to be the fault of other code removing our profile, and this is much better.

**Pull request merged**